specs/DesignDiscussions/DotNetTemplate.md

```diff
@@ -0,0 +1,348 @@
+    ],
+    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json"
+}
+```
+
+**Figure 6**  
+![](arrayType1.png)
+
+```json
+{
+    "type": "AdaptiveCard",
+    "version": "1.0",
+    "body": [
+        {
+            "type": "Container",
+            "$data": [
+                {
+                    "name": "Star",
+                    "nickname": "Dust"
+                },
+                {
+                    "name": "Death",
+                    "nickname": "Star"
+                }
+            ],
+            "items": [
+                {
+                    "type": "TextBlock",
+                    "text": "Header"
+                },
+                {
+                    "type": "FactSet",
+                    "facts": [
+                        {
+                            "title": "{name}",
+                            "value": "{nickname}"
+                        }
+                    ]
+                }
+            ]
+        }
+    ],
+    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json"
+}
+```  
+**Figure 7**  
+![](arrayType2.png)  
+
+Notice what's being repeated in **Figure 6** & **Figure 7**. In **Figure 6**, the FactSet is repeated since the data context is set at the FactSet. In **Figure 7**, the Container is repeated as the data context is sent at the Container.
+
+## Parsing Strategy
+It is obvious from the examples that simply returning the string or StringBuilder is not sufficient. We need ability to delay the expansion until the walker explores all of a subtree when its current data context is array, and one of its children makes reference to the data context.  
+
+### Result 
+- ANTLR allows Visitor to define a custom return object
+```dotnet
+public class JSONTemplateVisitor : JSONParserBaseVisitor<JSONTemplateVisitorResult>
+```
+- All of the Visitor methods will return this type once declared
+
+- .NET SDK will have JSONTemplateVisitorResult  
+- It has a linked list inside
+- It supports two operations, join and expand
+- join is used to merge results returned from other nodes
+ - when joining two results, if the joint (tail and head) are already expanded meaning no templated strings in them, then their linked list nodes are merged to one
+ - if they have templated string that needs to be expanded, then they are linked as linked lists without merging the nodes
+- In the VisitObj method, once visit to children finishes, we have one merged result object
+- if data context object was discovered in the current node, and it's array type, expansion happens with the merged result object
+ - it repeats the following steps n times where n is number of elements in data context
+ - it expands the result with data context. 
+ - the expanded result is merged
+- the result is returned from the method.
+- if data context object wasn't discovered in this node, simply return the current result without expanding, so it can be handled by a parent node which will have the data context
+
+## $when
```

Comment:
>$when [](start = 3, length = 5)

Discussion: Should this be $where like SQL? It's not really a filtering thing. This is really $showWhen or $showIf for a single object. Should we change to one of those? $when is consistent with Amazon's templating language (APL), is it valuable to match?

Reply:
Seems like we're agreed to stay with $when.

---
In reply to: [385959190](https://github.com/microsoft/AdaptiveCards/pull/3798#discussion_r385959190) [](ancestors = 385959190)